### ISSUE TITLE
Tunnel Source

### ISSUE THREAD

**Issue by motoridersd** [FingerlessGlov3s/OPNsensePIAWireguard#18]
How can one select what interface/WAN the wireguard traffic goes out of? I have two WAN connections and would like to make sure the Wireguard traffic only goes out one of them.

**Comment by FingerlessGlov3s** [comment]
Currently there's no way to select an outgoing interface for WireGuard on OPNsense. It'll look at the routing table and go from there, if you have no issue of swapping the default route for OPNsense to the one you want PIA connection to use, that should solve the issue for you. The other way to do it, is to create a static route and point it at the WAN of your choice, but as PIA server IPs will change this won't work in this case.

Sorry I can't offer a better ideal solution.

**Comment by motoridersd** [comment]
Ok... Is there a way to force the script to always use a specific list of servers? IE, I'm in Southern California and I only want to use the LA servers. Instead of adding ALL PIA hosts as static routes, I can add only the LA servers. This will work as long as all the static routes I added are valid (as long as PIA doesn't stop using all of them for example). I would need to make sure that the script will not use an IP that is not in my routing table statically pointed to the second WAN.

I might be able to add a script that will pull all la IPs from PIA every once in a while and update the routing table with the current IPs

**Comment by FingerlessGlov3s** [comment]
The script is built to only connect to a set region (like us_california) so this is already done, but far as I'm aware PIA don't provide a list of all servers in that region, the API they expose gets the best server for each region (based on some form of load metrics etc).

There is a routes API, although I'm not sure how reliable this would be. I would look at giving this a try but I personally don't have any Dual WAN setups.  If you can try adding the route manually in the interface and can prove its working, I can look to setting a Lab environment with Dual WAN and then look to implement this. Long as the API works as intended, it should be somewhat possible. Then I may be able to add a new variable at the time, one to specify which WAN you want to use. 

https://docs.opnsense.org/development/api/core/routes.html

**Comment by motoridersd** [comment]
I did try setting a static route for the active server and was able to force the traffic on the WAN port I wanted. So that works.

The addition you are thinking of would set the static route for the server that the API provided before it connects? That would be very useful. I'm happy to test that.

**Comment by FingerlessGlov3s** [comment]
Good to hear it'll work, yes that's the idea, when ever the script see's a requirement to reconnect, it'll add/change the static route, before it connects, therefore making the WireGuard connection to PIA use the specified WAN.

**Comment by FingerlessGlov3s** [comment]
Hello again, Can you try the below script, I've done some testing in my lab but when it comes to actual dual WAN, you may see some issues but hopefully it just works!!!  Please run it in debug mode first and remove your temporary route you added.

You will need to give the WireguardAPI User more access to modify the OPNsense routes. "System: Static Routes" should be the permission name your after.

https://paste.aes.pm/?6439d97fc8e6fdb5#7mErTiSQ1z3a5CmbYKRFFZ2VGfF1oZJPHSDdLQbDQ5RV

**Comment by FingerlessGlov3s** [comment]
I've done some more testing and now pushed the code in to the repo as commit https://github.com/FingerlessGlov3s/OPNsensePIAWireguard/commit/c7a41566cf97e9e5daa92a5a71abc7f9bcb4203c

Please download the latest script and look at the readme.
https://github.com/FingerlessGlov3s/OPNsensePIAWireguard

FYI: I have renamed `forcedWAN` to `tunnelGateway`, encase you were running that temporary script I posted above.

Please get back to me with your results, hopefully this will just work!

**Comment by motoridersd** [comment]
I just got around to testing and this seems to be working as expected. I see the static route was added and I can see the WG traffic going out the expected interface. Thank you!!

A couple of suggestions:

- The help doesn't mention that the name of the gateway interface must be in quotes. I got errors until I added those
- Is it possible to save the script parameters in a separate file? Updating the script requires adding the passwords and keys to the new file. If this info is saved elsewhere, ie, a file called PIAWireguardcredentials and referenced in the script, it would make updating a lot easier.
